### DISCUSSION TITLE
Ocarina of Time: Various Glide64 Graphical Glitches

### DISCUSSION THREAD

**Discussion by Millennium-Eagle** [mupen64plus-ae/mupen64plus-ae#924] [General]
Glide64 on this particular emulator massively cuts down on input lag on Ocarina of Time, but has several graphical glitches that I assume comes from the wrapper or something, because they don't occur on PC emulators with Glide64, and they occur across several different devices (Oneplus 8, Oneplus 3, Fire HD 8 Plus).

These shots come from Ocarina of Time randomizer, but these glitches also occur in vanilla Oot.
![Screenshot_20210715-013510_9784a853058524697766614cb28dab3d](https://user-images.githubusercontent.com/17091484/128002463-2eb7c51f-f09f-4467-b160-913405ffe14e.jpg)
The ground in this part of death mountain trail is always invisible unless you look at it from certain angles.
![Screenshot_20210715-235312_9784a853058524697766614cb28dab3d](https://user-images.githubusercontent.com/17091484/128002784-0e3d8270-1b1d-4c67-991d-092a7b77e283.jpg)
The Blue warp does not render correctly.
Various Menu graphics can become distorted at seemingly random.
![Screenshot_20210721-030127_9784a853058524697766614cb28dab3d](https://user-images.githubusercontent.com/17091484/128002964-04540e21-9341-45d0-99ba-4fc1201b9d49.jpg)
![Screenshot_20210721-030134_9784a853058524697766614cb28dab3d](https://user-images.githubusercontent.com/17091484/128003012-76922fdc-bf44-4afc-a324-00bc556a97e6.jpg)
Perhaps most severe is Link's graphics on the Equip screen replacing textures on various objects, characters, and geometry at seemingly random intervals.
![Screenshot_20210715-225254_9784a853058524697766614cb28dab3d](https://user-images.githubusercontent.com/17091484/128003363-aa0d0345-4020-41fb-91a3-44b636abebdb.jpg)
![Screenshot_20210729-225929_9784a853058524697766614cb28dab3d](https://user-images.githubusercontent.com/17091484/128003558-0f76d6ac-3397-4695-95d4-83f385e5a190.jpg)
![Screenshot_20210802-195309_9784a853058524697766614cb28dab3d](https://user-images.githubusercontent.com/17091484/128003599-f34d2ed1-bf1a-47e3-8619-d616f5770b7f.jpg)

**Comment by fzurita** [comment]
If you are having input lag with GLideN64, try disabling the threaded renderer under video settings. 

Glide64mk2 is not really maintained any more.

**Reply by Millennium-Eagle** [reply]
I do not use threaded renderer, but that does not remove the input lag. I believe its a problem with all n64 emulators and the frame buffer, which adds additional input lag that increases a lot with lower framerates.
The frame buffer IS supposed to add a small amount of input lag on console, but not to the degree that it does on current emulators. Glide64 gets around it by using smart frame buffer, I think, however for whatever reason Glide64 on your version of M64plus lowers the input lag significantly more than it does on PC emulators, where the difference is marginal. 
The difference is high enough on this emulator that it is worth using Glide64 with the graphical glitches for the reduced latency. It closely approaches Wii VC levels of responsiveness.

**Reply by fzurita** [reply]
Do you get the input lag with all other video plugins?

**Reply by Millennium-Eagle** [reply]
Yup, anything with frame buffer enabled, including Glide64 with Frame Buffer set to "Always On." 

To quote Gonetz on the issue,  "the problem is rather on emulator's side. The lag appears when graphics plugin uses VI origin register, which is updated by emu core. I can't see what I can do about it."

**Reply by fzurita** [reply]
Yeah, well, it's likely that the lag is in console as well. Does retroarch have the same lag with GLideN64?

Can you try selecting the game, going to settings, then advanced,  and adjusting count per operation to 1 as well as other parameters there to see if it has an effect?

**Reply by Millennium-Eagle** [reply]
The lag is not on console, I've tested it myself. Retroarch has the same problem on both its n64 cores; Parallel is actually slightly worse. Lowest lag in Retroarch comes from Mupen Core and GLideN64, where it matches what we have here. This is across both mobile and desktop varients. Editing various values in advanced does nothing across multiple emulators and platforms, including this one.

The input lag is harder to detect on other games, but is blatantly obvious in Ocarina of Time, due to its lower framerate, if you've played the Wii VC or N64 versions. For Oot speedrunners, the input lag make using emulators nearly unplayable for them, only using them for Oot Randos if they are playing Online with other people. Even for more casual players, its easy to see how much lag there is just by trying to aim with your bow or slingshot; its much, much harder than it should be. 

This is why this plugin is significant. For some reason, on this emulator in particular, the input lag is much lower than it usually is. I don't know if its the combination of a much more recent emulator with this plugin, or something in which the way the wrapper handles things, but it's much more responsive.

**Reply by fzurita** [reply]
OK, I see. Well, because Glide64mk2 doesn't have the lag it could be related to frame buffer emulation.

It could also be related to Android itself and disabling framebuffer emulation just makes the problem less obvious.

Try this, go to settings then shaders. Enable the CRT shader. That forces the Android UI to render at 60 fps rather than the game framerate because it's required for the shader to function correctly. If it's related to the Android UI, forcing 60fps updates may improve the problem.

**Reply by Millennium-Eagle** [reply]
Nope, toggling that didn't do a thing. In fact, I used a built in refresh rate monitor on my phone to see what it was, and it was 60 either way.

Although, funnily enough, it did reveal another oddity with either my phone or your emulator. When playing vertically, the screen does not change to 60hz when the game is loaded, and instead stays at 90hz like it usually is. Which is funny, and would be useful if this was a 30fps game; too bad it isn't.

For clarity, I play with wired and occasionally wireless gamepad, and the added lag is consistent across both setups, as well as when playing with touch screen controls.
